File: apps/easel/src/aic.rs
use rand::Rng;
use serde::Deserialize;
use std::time::Duration;

use crate::db::{self, DailyArtwork, Db};

const SEARCH_URL: &str = "https://api.artic.edu/api/v1/artworks/search";
const FIELDS: &str = "id,title,artist_display,artist_title,date_display,medium_display,dimensions,place_of_origin,credit_line,description,short_description,image_id";

pub fn build_client() -> reqwest::Client {
    reqwest::Client::builder()
        .timeout(Duration::from_secs(20))
        .user_agent("andromeda-easel/0.1 (+https://github.com/stevedylandev/andromeda)")
        .build()
        .expect("Failed to build HTTP client")
}

#[derive(Debug, Deserialize)]
pub struct RawArtwork {
    pub id: i64,
    pub title: Option<String>,
    pub artist_display: Option<String>,
    pub artist_title: Option<String>,
    pub date_display: Option<String>,
    pub medium_display: Option<String>,
    pub dimensions: Option<String>,
    pub place_of_origin: Option<String>,
    pub credit_line: Option<String>,
    pub description: Option<String>,
    pub short_description: Option<String>,
    pub image_id: Option<String>,
}

#[derive(Debug, Deserialize)]
struct Pagination {
    total: u64,
}

#[derive(Debug, Deserialize)]
struct SearchResponse<T> {
    pagination: Pagination,
    data: Vec<T>,
}

#[derive(Debug, Deserialize)]
struct IdOnly {
    #[allow(dead_code)]
    id: i64,
}

fn build_params(classifications: &[String]) -> String {
    let terms: Vec<serde_json::Value> = classifications
        .iter()
        .map(|c| serde_json::Value::String(c.to_lowercase()))
        .collect();
    let body = serde_json::json!({
        "query": {
            "bool": {
                "must": [
                    { "term": { "is_public_domain": true } },
                    { "terms": { "classification_title.keyword": terms } },
                    { "exists": { "field": "image_id" } }
                ]
            }
        }
    });
    body.to_string()
}

pub async fn total_matching(
    client: &reqwest::Client,
    classifications: &[String],
) -> Result<u64, String> {
    let params = build_params(classifications);
    let url = format!(
        "{SEARCH_URL}?params={}&limit=1&fields=id",
        urlencoding::encode(&params)
    );
    let resp = client
        .get(&url)
        .send()
        .await
        .map_err(|e| format!("count fetch failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(format!("count returned status {}", resp.status()));
    }
    let body: SearchResponse<IdOnly> = resp
        .json()
        .await
        .map_err(|e| format!("count parse failed: {e}"))?;
    Ok(body.pagination.total)
}

pub async fn fetch_artwork_at(
    client: &reqwest::Client,
    classifications: &[String],
    page: u64,
) -> Result<Option<RawArtwork>, String> {
    let params = build_params(classifications);
    let url = format!(
        "{SEARCH_URL}?params={}&limit=1&page={page}&fields={FIELDS}",
        urlencoding::encode(&params)
    );
    let resp = client
        .get(&url)
        .send()
        .await
        .map_err(|e| format!("artwork fetch failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(format!("artwork returned status {}", resp.status()));
    }
    let mut body: SearchResponse<RawArtwork> = resp
        .json()
        .await
        .map_err(|e| format!("artwork parse failed: {e}"))?;
    Ok(body.data.pop())
}

pub async fn pick_unique(
    client: &reqwest::Client,
    db: &Db,
    classifications: &[String],
    max_retries: u32,
) -> Result<RawArtwork, String> {
    let total = total_matching(client, classifications).await?;
    if total == 0 {
        return Err("AIC search returned zero matches for given classifications".to_string());
    }

    for attempt in 0..=max_retries {
        let page = {
            let mut rng = rand::thread_rng();
            rng.gen_range(1..=total)
        };
        let art = match fetch_artwork_at(client, classifications, page).await? {
            Some(a) => a,
            None => continue,
        };
        if art.image_id.is_none() || art.image_id.as_deref() == Some("") {
            tracing::warn!("artwork {} has no image_id, retrying", art.id);
            continue;
        }
        match db::artwork_id_exists(db, art.id) {
            Ok(true) => {
                tracing::info!(
                    "duplicate artwork {} on attempt {}, retrying",
                    art.id,
                    attempt + 1
                );
                continue;
            }
            Ok(false) => return Ok(art),
            Err(e) => return Err(format!("dedup check failed: {e}")),
        }
    }
    Err(format!(
        "failed to pick a non-duplicate artwork after {} retries",
        max_retries + 1
    ))
}

pub fn raw_to_daily(raw: RawArtwork, date: String, fetched_at: String) -> Option<DailyArtwork> {
    let image_id = raw.image_id?;
    if image_id.is_empty() {
        return None;
    }
    Some(DailyArtwork {
        date,
        artwork_id: raw.id,
        title: raw.title.unwrap_or_else(|| "Untitled".to_string()),
        artist_display: raw.artist_display,
        artist_title: raw.artist_title,
        date_display: raw.date_display,
        medium_display: raw.medium_display,
        dimensions: raw.dimensions,
        place_of_origin: raw.place_of_origin,
        credit_line: raw.credit_line,
        description: raw.description,
        short_description: raw.short_description,
        image_id,
        fetched_at,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_params_lowercases_classifications() {
        let p = build_params(&["Painting".to_string(), "DRAWING".to_string()]);
        assert!(p.contains("\"painting\""));
        assert!(p.contains("\"drawing\""));
        assert!(p.contains("is_public_domain"));
        assert!(p.contains("image_id"));
    }
}

File: apps/easel/aic.go
package main

import (
	"bytes"
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"math/rand/v2"
	"net/http"
	"net/url"
	"time"
)

const aicSearchURL = "https://api.artic.edu/api/v1/artworks/search"

var aicFields = []string{
	"id", "title", "artist_display", "artist_title", "date_display",
	"medium_display", "dimensions", "place_of_origin", "credit_line",
	"description", "short_description", "image_id",
}

var aicExcludeFields = []string{
	"title", "description", "short_description", "term_titles", "subject_titles",
	"category_titles", "classification_titles",
}

type rawArtwork struct {
	ID               int64   `json:"id"`
	Title            *string `json:"title"`
	ArtistDisplay    *string `json:"artist_display"`
	ArtistTitle      *string `json:"artist_title"`
	DateDisplay      *string `json:"date_display"`
	MediumDisplay    *string `json:"medium_display"`
	Dimensions       *string `json:"dimensions"`
	PlaceOfOrigin    *string `json:"place_of_origin"`
	CreditLine       *string `json:"credit_line"`
	Description      *string `json:"description"`
	ShortDescription *string `json:"short_description"`
	ImageID          *string `json:"image_id"`
}

type searchResponse struct {
	Pagination struct {
		Total uint64 `json:"total"`
	} `json:"pagination"`
	Data []rawArtwork `json:"data"`
}

func buildHTTPClient() *http.Client {
	return &http.Client{Timeout: 20 * time.Second}
}

func buildAICParams(classifications, excludeTerms []string) string {
	terms := make([]string, 0, len(classifications))
	for _, c := range classifications {
		terms = append(terms, lower(c))
	}
	mustNot := make([]map[string]any, 0, len(excludeTerms))
	for _, t := range excludeTerms {
		mustNot = append(mustNot, map[string]any{
			"multi_match": map[string]any{
				"query":  t,
				"fields": aicExcludeFields,
				"type":   "phrase",
			},
		})
	}
	body := map[string]any{
		"query": map[string]any{
			"bool": map[string]any{
				"must": []any{
					map[string]any{"term": map[string]any{"is_public_domain": true}},
					map[string]any{"terms": map[string]any{"classification_title.keyword": terms}},
					map[string]any{"exists": map[string]any{"field": "image_id"}},
				},
				"must_not": mustNot,
			},
		},
	}
	buf, _ := json.Marshal(body)
	return string(buf)
}

func lower(s string) string {
	b := []byte(s)
	for i, c := range b {
		if c >= 'A' && c <= 'Z' {
			b[i] = c + 32
		}
	}
	return string(b)
}

func aicTotalMatching(ctx context.Context, client *http.Client, classifications, excludeTerms []string) (uint64, error) {
	params := buildAICParams(classifications, excludeTerms)
	u := fmt.Sprintf("%s?params=%s&limit=1&fields=id", aicSearchURL, url.QueryEscape(params))
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return 0, err
	}
	req.Header.Set("User-Agent", "andromeda-easel/0.1 (+https://github.com/stevedylandev/andromeda)")
	resp, err := client.Do(req)
	if err != nil {
		return 0, fmt.Errorf("count fetch failed: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return 0, fmt.Errorf("count status %s", resp.Status)
	}
	var sr searchResponse
	if err := json.NewDecoder(resp.Body).Decode(&sr); err != nil {
		return 0, err
	}
	return sr.Pagination.Total, nil
}

func aicFetchArtworkAt(ctx context.Context, client *http.Client, classifications, excludeTerms []string, page uint64) (*rawArtwork, error) {
	params := buildAICParams(classifications, excludeTerms)
	u := fmt.Sprintf("%s?params=%s&limit=1&page=%d&fields=%s",
		aicSearchURL, url.QueryEscape(params), page, joinFields(aicFields))
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "andromeda-easel/0.1")
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("artwork fetch failed: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("artwork status %s", resp.Status)
	}
	body := &bytes.Buffer{}
	if _, err := body.ReadFrom(resp.Body); err != nil {
		return nil, err
	}
	var sr searchResponse
	if err := json.Unmarshal(body.Bytes(), &sr); err != nil {
		return nil, err
	}
	if len(sr.Data) == 0 {
		return nil, nil
	}
	return &sr.Data[len(sr.Data)-1], nil
}

func joinFields(fs []string) string {
	out := ""
	for i, f := range fs {
		if i > 0 {
			out += ","
		}
		out += f
	}
	return out
}

func pickUnique(ctx context.Context, client *http.Client, db *sql.DB, classifications, excludeTerms []string, maxRetries int) (*rawArtwork, error) {
	total, err := aicTotalMatching(ctx, client, classifications, excludeTerms)
	if err != nil {
		return nil, err
	}
	if total == 0 {
		return nil, fmt.Errorf("AIC search returned zero matches")
	}
	for attempt := 0; attempt <= maxRetries; attempt++ {
		page := rand.Uint64N(total) + 1
		art, err := aicFetchArtworkAt(ctx, client, classifications, excludeTerms, page)
		if err != nil {
			return nil, err
		}
		if art == nil || art.ImageID == nil || *art.ImageID == "" {
			continue
		}
		exists, err := artworkIDExists(db, art.ID)
		if err != nil {
			return nil, fmt.Errorf("dedup check: %w", err)
		}
		if exists {
			continue
		}
		return art, nil
	}
	return nil, fmt.Errorf("failed to pick non-duplicate artwork after %d retries", maxRetries+1)
}

func rawToDaily(r *rawArtwork, date, fetchedAt string) *DailyArtwork {
	if r.ImageID == nil || *r.ImageID == "" {
		return nil
	}
	title := "Untitled"
	if r.Title != nil && *r.Title != "" {
		title = *r.Title
	}
	d := &DailyArtwork{
		Date:      date,
		ArtworkID: r.ID,
		Title:     title,
		ImageID:   *r.ImageID,
		FetchedAt: fetchedAt,
	}
	if r.ArtistDisplay != nil {
		d.ArtistDisplay = sql.NullString{String: *r.ArtistDisplay, Valid: true}
	}
	if r.ArtistTitle != nil {
		d.ArtistTitle = sql.NullString{String: *r.ArtistTitle, Valid: true}
	}
	if r.DateDisplay != nil {
		d.DateDisplay = sql.NullString{String: *r.DateDisplay, Valid: true}
	}
	if r.MediumDisplay != nil {
		d.MediumDisplay = sql.NullString{String: *r.MediumDisplay, Valid: true}
	}
	if r.Dimensions != nil {
		d.Dimensions = sql.NullString{String: *r.Dimensions, Valid: true}
	}
	if r.PlaceOfOrigin != nil {
		d.PlaceOfOrigin = sql.NullString{String: *r.PlaceOfOrigin, Valid: true}
	}
	if r.CreditLine != nil {
		d.CreditLine = sql.NullString{String: *r.CreditLine, Valid: true}
	}
	if r.Description != nil {
		d.Description = sql.NullString{String: *r.Description, Valid: true}
	}
	if r.ShortDescription != nil {
		d.ShortDescription = sql.NullString{String: *r.ShortDescription, Valid: true}
	}
	return d
}
